examples/render/quad_render/main.rs

```diff
@@ -280,56 +285,30 @@ fn main() {
         w: pixel_width, h: pixel_height,
     };
 
-    let mut fence = device.create_fence(false);
-    let mut graphics_pool = context.mut_queue()
-        .create_graphics_pool(16, pool::CommandPoolCreateFlags::empty());
-
-    println!("copy buffer to texture");
-    {
-        let submit = {
-            let mut cmd_buffer = graphics_pool.acquire_command_buffer();
-
-            let image_barrier = m::Barrier::Image {
-                states: (i::Access::empty(), i::ImageLayout::Undefined) ..
-                        (i::TRANSFER_WRITE, i::ImageLayout::TransferDstOptimal),
-                target: image.resource(),
-                range: (0..1, 0..1),
-            };
-            cmd_buffer.pipeline_barrier(pso::TOP_OF_PIPE .. pso::TRANSFER, &[image_barrier]);
-
-            cmd_buffer.copy_buffer_to_image(
-                image_upload_buffer.resource(),
-                image.resource(),
-                i::ImageLayout::TransferDstOptimal,
-                &[command::BufferImageCopy {
-                    buffer_offset: 0,
-                    buffer_row_pitch: row_pitch,
-                    buffer_slice_pitch: row_pitch * (height as u32),
-                    image_aspect: i::ASPECT_COLOR,
-                    image_subresource: (0, 0..1),
-                    image_offset: command::Offset { x: 0, y: 0, z: 0 },
-                    image_extent: d::Extent { width, height, depth: 1 },
-                }]);
-
-            let image_barrier = m::Barrier::Image {
-                states: (i::TRANSFER_WRITE, i::ImageLayout::TransferDstOptimal) ..
-                        (i::SHADER_READ, i::ImageLayout::ShaderReadOnlyOptimal),
-                target: image.resource(),
-                range: (0..1, 0..1),
-            };
-            cmd_buffer.pipeline_barrier(pso::TRANSFER .. pso::BOTTOM_OF_PIPE, &[image_barrier]);
-
-            cmd_buffer.finish()
-        };
-
-        let submission = Submission::new()
-            .submit(&[submit]);
-        context.mut_queue().submit(submission, Some(&mut fence));
-
-        device.wait_for_fences(&[&fence], d::WaitFor::All, !0);
-    }
-
-    device.destroy_fence(fence);
+    init_encoder.copy_buffer_to_image(
+        &image_upload_buffer,
+        &image,
+        &[command::BufferImageCopy {
+            buffer_offset: 0,
+            buffer_row_pitch: row_pitch,
+            buffer_slice_pitch: row_pitch * (height as u32),
+            image_aspect: i::ASPECT_COLOR,
+            image_subresource: (0, 0..1),
+            image_offset: command::Offset { x: 0, y: 0, z: 0 },
+            image_extent: d::Extent { width, height, depth: 1 },
+        }]);
+
+    let init_submit = init_encoder.finish();
```

Comment:
That's what I'm worried about: passing the encoder in the device creation doesn't make it safe: user is still responsible to actually submit it. So I think we should change to tokens right away:
  - resource creation returns a token
  - encoder can do `init_resources(&[Token])`, which consumes the tokens.

This is not safer but I believe is more universal.

Reply:
That's also better for batching into a single barrier command :)